### PR TITLE
Fix Issue 23957 - Casting to derived extern(C++) class is unsafe

### PR DESCRIPTION
**Pull request by ntrel**
So deprecate those in safe code.

### CODE REVIEW

**Comment by ntrel** [comment]
Closing in favour of #15293.

**Comment by thewilsonator** [comment]
are you intending for this to be merged (instead of https://github.com/dlang/dmd/pull/15293)?

**Comment by ntrel** [comment]
@thewilsonator Yes, I think it would be better just to update the spec and make the casts not `@safe` (see https://github.com/dlang/dmd/pull/15293#issuecomment-2158258554).

File: compiler/src/dmd/safe.d [hunk 1]
```diff
@@ -182,8 +184,11 @@ bool isSafeCast(Expression e, Type tfrom, Type tto)
                 && cdfrom.classKind == ClassKind.d && cdto.classKind == ClassKind.d))
             return false;
 
+        // no RTTI
         if (cdfrom.isCPPinterface() || cdto.isCPPinterface())
             return false;
+        if (cdfrom.classKind == ClassKind.cpp || cdto.classKind == ClassKind.cpp)
+            return -1;
```
Comment (thewilsonator) [inline compiler/src/dmd/safe.d:191]: we should not mix `return false;` and `return -1;` in the same function.

File: compiler/src/dmd/expressionsem.d [hunk 2]
```diff
@@ -9251,13 +9251,14 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
-        if (!isSafeCast(ex, t1b, tob))
-        {
-            if (sc.setUnsafe(false, exp.loc, "cast from `%s` to `%s` not allowed in safe code", exp.e1.type, exp.to))
-            {
-                return setError();
-            }
-        }
+        auto r = isSafeCast(ex, t1b, tob);
+        if (r == -1 && sc.setUnsafePreview(FeatureState.default_, false, exp.loc,
+                "cast from `%s` to `%s` not allowed in safe code", exp.e1.type, exp.to))
```
Comment (thewilsonator) [inline compiler/src/dmd/expressionsem.d:9256]: If I understand correctly, it is downcasts are unsafe? If so, can you make the error message convey that?

**Review by thewilsonator — COMMENTED** [review]
Otherwise looks good

**Review by thewilsonator — APPROVED** [review]

**Pull request merged**